File: examples/02_request_response.py
"""
Request-Response - Proxy Tell vs Proxy Ask Patterns
====================================================

This example demonstrates two ways to call agent methods from outside the actor system:

- **proxy_tell (fire-and-forget):** Calls an agent method without waiting for a return
  value. The method sends a message to another agent, which replies asynchronously.
- **proxy_ask (blocking):** Calls an agent method and blocks until the method returns.
  The caller receives the return value directly — no inter-agent message is sent.

Key types used:
- CalculationRequest / CalculationResult: request-response message pair (tell path)
- proxy_tell / proxy_ask: two proxy modes on ActorSystem

Run with:  python examples/02_request_response.py
Or with:  uv run python examples/02_request_response.py
"""

from __future__ import annotations

import time
import uuid

from akgentic.core import ActorAddress, ActorSystem, Akgent, BaseConfig, BaseState
from akgentic.core.messages import Message

# =============================================================================
# STEP 1: Define request and response message types
# =============================================================================
# Messages are the way agents communicate. Every message extends Message,
# which provides automatic serialization, unique IDs, and sender tracking.


class CalculationRequest(Message):
    """A calculation request message sent to the calculator agent.

    Attributes:
        a: First operand (integer).
        b: Second operand (integer).
        operation: The operation to perform ('+', '-', '*', '/').
    """

    a: int
    b: int
    operation: str


class CalculationResult(Message):
    """A calculation result message sent back from the calculator agent.

    Attributes:
        result: The computed result of the calculation.
        request_id: The ID of the original request message.
    """

    result: float
    request_id: uuid.UUID


# =============================================================================
# STEP 2: Define the calculator agent that processes requests
# =============================================================================
# This agent handles incoming CalculationRequest messages and sends back
# CalculationResult messages with the computed values.


class CalculatorAgent(Akgent[BaseConfig, BaseState]):
    """An agent that processes calculation requests and sends results.

    This agent demonstrates:
    - Receiving request messages (CalculationRequest)
    - Processing the request
    - Sending response messages back to the requester
    - The receiveMsg_<Type> pattern for automatic message dispatch
    """

    def receiveMsg_CalculationRequest(
        self, message: CalculationRequest, sender: ActorAddress
    ) -> None:
        """Handle incoming calculation request by computing and sending result.

        Args:
            message: The CalculationRequest containing operands and operation.
            sender: The ActorAddress of the sender (ClientAgent in this case).
        """
        # Extract values from request
        a = message.a
        b = message.b
        operation = message.operation

        # Compute result based on operation
        print(f"[CalculatorAgent] Processing request: {a} {operation} {b}")
        result: float
        if operation == "+":
            result = float(a + b)
        elif operation == "-":
            result = float(a - b)
        elif operation == "*":
            result = float(a * b)
        elif operation == "/":
            result = a / b if b != 0 else float("nan")
        else:
            result = float("nan")

        # Send result back to the sender
        self.send(
            sender,
            CalculationResult(result=result, request_id=message.id),
        )


# =============================================================================
# STEP 3: Define the client agent that sends requests
# =============================================================================
# This agent demonstrates both fire-and-forget (tell) and blocking (ask) patterns.


class ClientAgent(Akgent[BaseConfig, BaseState]):
    """An agent that sends calculation requests and receives results.

    This agent demonstrates:
    - Using tell() for fire-and-forget message sending
    - Using ask() for blocking/synchronous communication
    - Receiving response messages (CalculationResult)
    """

    def send_request_tell(
        self, calculator_addr: ActorAddress, a: int, b: int, operation: str
    ) -> None:
        """Send a calculation request using tell (fire-and-forget).

        tell() sends the message but does NOT wait for a response.
        This is useful for asynchronous, decoupled communication where you
        don't need immediate feedback from the recipient.

        Args:
            calculator_addr: Address of the calculator agent.
            a: First operand.
            b: Second operand.
            operation: Operation to perform.
        """
        print(f"[ClientAgent] Sending calculation request: {a} {operation} {b}")
        request = CalculationRequest(a=a, b=b, operation=operation)
        # send() internally uses tell - message is sent but we don't wait
        self.send(calculator_addr, request)

    def send_request_ask(
        self, _calculator_addr: ActorAddress, a: int, b: int, operation: str
    ) -> dict[str, float | uuid.UUID]:
        """Compute a calculation locally and return the result (ask pattern).

        Unlike send_request_tell, this method does NOT send a message to the
        CalculatorAgent. It computes the result directly and returns it.
        When called via proxy_ask, the caller blocks until this method returns
        and receives the return value — demonstrating the blocking proxy pattern.

        The arithmetic logic is intentionally duplicated from CalculatorAgent
        to keep each path self-contained for demonstration purposes.

        Args:
            _calculator_addr: Unused — kept for API symmetry with
                send_request_tell.
            a: First operand.
            b: Second operand.
            operation: Operation to perform.

        Returns:
            Dictionary with the computed result and a standalone request ID
            (not correlated to any message, since no message is sent).
        """
        print(f"[ClientAgent] Computing calculation locally (ask): {a} {operation} {b}")
        # No message is sent to another agent — the computation happens right
        # here inside the ClientAgent. The proxy_ask caller will block until
        # this method returns, then receive the return value directly.
        # (Arithmetic is duplicated from CalculatorAgent intentionally.)
        result: float
        if operation == "+":
            result = float(a + b)
        elif operation == "-":
            result = float(a - b)
        elif operation == "*":
            result = float(a * b)
        elif operation == "/":
            result = a / b if b != 0 else float("nan")
        else:
            result = float("nan")

        # request_id is an arbitrary UUID here (no correlated request message).
        return {"result": result, "request_id": uuid.uuid4()}

    def receiveMsg_CalculationResult(
        self, message: CalculationResult, sender: ActorAddress
    ) -> None:
        """Handle incoming calculation result.

        Args:
            message: The CalculationResult containing the computed value.
            sender: The ActorAddress of the sender (CalculatorAgent).
        """
        print(f"[ClientAgent] Received result: {message.result}")


# =============================================================================
# STEP 4: Main execution - create system, agents, and run
# =============================================================================


def main() -> None:
    """Run the Request-Response example demonstrating proxy_tell and proxy_ask."""
    print("[Request-Response] Demonstrating proxy_tell and proxy_ask patterns...")

    # Create the actor system - this is the runtime that manages all agents
    # ActorSystem provides zero-dependency local execution (no Redis, etc.)
    actor_system = ActorSystem()

    try:
        # Create the CalculatorAgent first - it needs to exist before ClientAgent
        # can send requests to it
        calculator_addr = actor_system.createActor(
            CalculatorAgent,
            config=BaseConfig(name="calculator", role="Calculator"),
        )

        # Create the ClientAgent
        client_addr = actor_system.createActor(
            ClientAgent,
            config=BaseConfig(name="client", role="Client"),
        )

        # Get a proxy to the client agent using tell pattern
        # This allows us to call methods on the client without waiting for responses
        client_tell = actor_system.proxy_tell(client_addr, ClientAgent)

        # Demonstrate tell pattern (fire-and-forget)
        # The client sends a request but doesn't wait for the response
        client_tell.send_request_tell(calculator_addr, 10, 5, "+")

        # Wait for the async message to be processed
        time.sleep(0.2)

        # Get a proxy to the client agent using ask pattern.
        # proxy_ask blocks until the called method returns and gives back its
        # return value — no inter-agent message is involved in this path.
        client_ask = actor_system.proxy_ask(client_addr, ClientAgent)

        # Demonstrate ask pattern (blocking).
        # send_request_ask computes locally inside ClientAgent and returns the
        # result. proxy_ask blocks here until the method finishes, then we get
        # the return value directly — unlike tell, no CalculatorAgent is used.
        result = client_ask.send_request_ask(calculator_addr, 20, 3, "*")
        print(f"[Request-Response] Ask result: {result}")

        print("[Request-Response] All calculations complete. Shutting down.")

    finally:
        # Always clean up the actor system
        actor_system.shutdown(timeout=5)


if __name__ == "__main__":
    main()

File: examples/02_request_response.exs
# Request/Response Patterns — Blocking vs Fire-and-Forget
#
# This example demonstrates two communication patterns between agents:
#
# 1. Fire-and-forget (async): `Akgentic.signal_async/3`
#    Maps to Python's `proxy_tell()` — send a signal and continue immediately.
#
# 2. Blocking request-response (sync): `Akgentic.signal/3`
#    Maps to Python's `proxy_ask()` — send a signal and wait for the result.
#
# Run with: mix run examples/02_request_response.exs

# =============================================================================
# STEP 1: Define actions
# =============================================================================

defmodule Examples.RequestResponse.Actions.Add do
  @moduledoc "Adds two numbers and stores the result."
  use Jido.Action,
    name: "add",
    description: "Adds two numbers",
    schema: [
      a: [type: :float, required: true],
      b: [type: :float, required: true]
    ]

  def run(%{a: a, b: b}, _context) do
    result = a + b
    IO.puts("[CalculatorAgent] add(#{a}, #{b}) = #{result}")
    {:ok, %{last_result: result, last_operation: "add"}}
  end
end

defmodule Examples.RequestResponse.Actions.Multiply do
  @moduledoc "Multiplies two numbers and stores the result."
  use Jido.Action,
    name: "multiply",
    description: "Multiplies two numbers",
    schema: [
      a: [type: :float, required: true],
      b: [type: :float, required: true]
    ]

  def run(%{a: a, b: b}, _context) do
    result = a * b
    IO.puts("[CalculatorAgent] multiply(#{a}, #{b}) = #{result}")
    {:ok, %{last_result: result, last_operation: "multiply"}}
  end
end

# =============================================================================
# STEP 2: Define the CalculatorAgent
# =============================================================================

defmodule Examples.RequestResponse.CalculatorAgent do
  @moduledoc "An agent that performs arithmetic operations."
  use Akgentic.Agent,
    name: "calculator",
    description: "Performs arithmetic operations",
    schema: [
      last_result: [type: :float, default: 0.0],
      last_operation: [type: :string, default: ""]
    ],
    actions: [
      Examples.RequestResponse.Actions.Add,
      Examples.RequestResponse.Actions.Multiply
    ],
    signal_routes: [
      {"add", Examples.RequestResponse.Actions.Add},
      {"multiply", Examples.RequestResponse.Actions.Multiply}
    ]
end

# =============================================================================
# STEP 3: Pure functional path — cmd/2 (synchronous, no OTP process needed)
# =============================================================================

IO.puts("\n=== Pure Functional Path (cmd/2) ===\n")

alias Examples.RequestResponse.Actions.Add
alias Examples.RequestResponse.Actions.Multiply
alias Examples.RequestResponse.CalculatorAgent

agent = CalculatorAgent.new()
IO.puts("Initial state: last_result=#{agent.state.last_result}")

{agent, _directives} = CalculatorAgent.cmd(agent, {Add, %{a: 10.0, b: 5.0}})
IO.puts("After add(10, 5): last_result=#{agent.state.last_result}")

{agent, _directives} = CalculatorAgent.cmd(agent, {Multiply, %{a: 4.0, b: 7.0}})
IO.puts("After multiply(4, 7): last_result=#{agent.state.last_result}")

# =============================================================================
# STEP 4: OTP path — blocking (signal/3) vs fire-and-forget (signal_async/3)
# =============================================================================

IO.puts("\n=== OTP Path: Blocking (signal/3) ===\n")

# Start the agent under the OTP supervisor
{:ok, pid} = Akgentic.start_agent(CalculatorAgent, id: "calc-1")

# Blocking call — equivalent to Python's proxy_ask()
# Returns {:ok, agent} so we can inspect the result immediately.
{:ok, result_agent} = Akgentic.signal(pid, "add", %{a: 3.0, b: 7.0})
IO.puts("Blocking add(3, 7) result: #{result_agent.state.last_result}")

{:ok, result_agent} = Akgentic.signal(pid, "multiply", %{a: 6.0, b: 9.0})
IO.puts("Blocking multiply(6, 9) result: #{result_agent.state.last_result}")

IO.puts("\n=== OTP Path: Fire-and-Forget (signal_async/3) ===\n")

# Non-blocking cast — equivalent to Python's proxy_tell()
# Returns :ok immediately; the agent processes the signal asynchronously.
:ok = Akgentic.signal_async(pid, "add", %{a: 100.0, b: 200.0})
IO.puts("Fire-and-forget add(100, 200) dispatched — continuing immediately")

# Give the agent time to process the async signal
Process.sleep(200)

# Confirm result with a blocking call
{:ok, confirmed} = Akgentic.signal(pid, "add", %{a: 0.0, b: 0.0})
IO.puts("Previous async operation was: #{confirmed.state.last_operation}")

Akgentic.stop_agent(pid)
IO.puts("\n[Request/Response] Done!")
